keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,425 @@
+Supporting multiple ResourceClass matching also makes it easy to ensure that existing pods requesting resources through raw resource name can continue to be scheduled properly when administrators add ResourceClass in a cluster. To guarantee this, the scheduler may just consider raw resource as a special ResourceClass with empty resource metadata constraints.
+
+Because a ComputeResource can match multiple ResourceClasses, Scheduler and Kubelet need to ensure a consistent view on ComputeResource to ResourceClass request binding. Let us consider an example to illustrate this problem. Suppose a node has two ComputeResources, CR1 and CR2, that have the same raw resource name but different sets of properties. Suppose they both satisfy the property constraints of ResourceClass RC1, but only CR2 satisfies the property constraints of another ResourceClass RC2. Suppose a Pod requesting RC1 is scheduled first. Because the RC1 resource request can be satisfied by either CR1 or CR2, it is important for the scheduler to record the binding information and propagate it to Kubelet, and Kubelet should honor this binding instead of making its own binding decision. This way, when another Pod comes in that requests RC2, the scheduler can determine whether Pod can fit on the node or not, depending on whether the previous RC1 request is bound to CR1 or CR2.
+
+To maintain and propagate ResourceClass to ComputeResource binding information, the scheduler will need to record this information in a newly introduced ContainerSpec field, similar to the existing NodeName field, and Kubelet will need to consume this information. During the initial implementation, we propose to encode the ResourceClass to the underlying compute resource binding information in a new `AllocatedDeviceIDs map[v1.ResourceName][]types.UID` field in ContainerSpec. Adding this field has been discussed as a possible solution to support other use cases, such as third-party resource monitoring and network device plugins. For the purpose to support ResourceClass, we will extend the scheduler NodeInfo cache to store ResourceClass to the matching ComputeResource information on the node. For a given ComputeResource, its capacity will be reflected in NodeInfo.allocatableResource with all matching ResourceClass names. This way, the current node resource fitness evaluation will stay most the same. After a pod is bound to a node, the scheduler will choose the requested number of devices from the matching ComputeResource on the node, and record this information in the mentioned new field. After that, it increases the NodeInfo.requestedResource for all of the matching ResourceClass names of that ComputeResource. Note that if the AllocatedDeviceIDs field is pre-specified, scheduler should honor this binding instead of overwriting it, similar to how it handles pre-specified NodeName.
```

Comment:
This seems to hoist some node-level topology concerns into the scheduler. Is that intentional? So far we have tried really hard to avoid doing that.

Is it feasible to instead record the matching `ComputeResource` name, and not the device ID? It seems that all of the device IDs within a ComputeResource category on a node are intended to be fungible for scheduling purposes. Anyway, they have identical properties. This would leave some flexibility for the node to continue to do late-binding of deviceIDs.

If native resources are later migrated to this API, this field also seems to allow manual CPU scheduling from the container spec, which we've discussed and rejected a few times in the past.

Reply:
Not sure about the encoding either. I think the most important part to let the scheduler only worry about mapping resource class requests to concrete compute resources. From that point on, publishing the device ID bindings done by the node via the API would be an optional extension. No such advertisement currently happens, and I don't think it would be needed to enable the basic flow outlined in this proposal.

Reply:
Two guiding principles (in my mind at least):
* the scheduler should not know about device IDs
* the node should not know about resource classes

Reply:
Something like the above may be desirable anyhow, in order to facilitate a split between node and scheduler feasibility checks. Right now the node uses the same predicates as the scheduler, however there's no way in this proposal to disambiguate between resource class names and concrete resource names. If we let the scheduler sort all that out and write only concrete names into this new spec, it could greatly simplify the pod admission logic on the node.

Reply:
thanks and agree with all that you said above. @dashpole can we align this i.e `AllocatedComputeResources` with your node monitoring proposal where you are suggesting to add `DeviceIds` in the container spec. May be we can have `AllocatedDeviceIds` along with `Quantity`, which would be filled by scheduler and kubelet will populate `IDs`. WDYT?
/cc @jiayingz

Reply:
As mentioned on the google doc, we should get the design of this PR nailed down first. I don't think this is the case now.
I think the current shape of the document to add Device ID to the Pod API object is good and should unblock monitoring. Resource Class should not be a blocker for that.

@ConnorDoyle you mentioned the guiding principle of `the scheduler should not know about device IDs`, I believe we should at least keep the flexibility for that to be possible in another scheduler.

It also seems reasonable to me to adopt the same as PVCs rather than fork our own.

Reply:
I think the overall concept of scheduler doing ResourceClass -> ComputeResource binding, and then kubelet doing ComputeResource -> Device binding makes sense.  I'll think some more about how best to represent it in a go struct (maybe seperate items?  `AllocatedComputeResources` and `AssignedDevices`?).

I am curious how the multi-device story will evolve.  With the ability for a custom scheduler to assign devices, we could have an out-of-tree solution for this.

Reply:
@dashpole one argument for separate structures is that the device ID field would only be used for compute resources exported by a device plugin.

@RenaudWasTaken allowing the scheduler to fill in concrete device bindings might be OK. That decision should be made with knowledge of the prior discussions (thinking mostly of @thockin's comments in https://github.com/kubernetes/community/pull/171).

Reply:
Sorry for joining the discussions late. I was ooo last week.

I think there are a few reasons we propose to have scheduler encode ResourceClass -> ComputeResource binding in the current format. 1) Minimize API changes during initial phase. As mentioned, there has been a separate proposal to recording assigned device information in ContainerSpec to support monitoring and maybe also network device plugins. We propose to reuse that field to also encode ResourceClass to ComputeResource binding, as we want to be careful on introducing new API fields in ContainerSpec. 2) To simplify ComputeResource life cycle during initial phase. In the original google doc, we proposed to include a Name field in ComputeResource struct that uniquely represents the compute resources with unique properties on a node, and have scheduler encode ResourceClass->ComputeResource binding info through that name. The complicated part of that approach is that scheduler and kubelet can have inconsistent allocation views when a device changes its properties. We have discussed a few ways to handle the situation, but I think each of them involves certain complexity on kubelet or during upgrade. Therefore we propose to have scheduler directly encode what device is allocated to a Container during the initial phase, which I think allows simpler compute resource life cycle management. 3) For topology aware scheduling, quite likely, I think we may need to enable both scheduler level guaranteed topology aware scheduling and node level best effort topology aware scheduling, hopefully through extensions. This proposal addresses the first missing part. Agree with @ConnorDoyle that we should clarify how the current proposal can be extended later to also allow node level best effort topology aware scheduling. I like the idea @dashpole suggested that we may use AllocatedComputeResources struct that currently only contains AssignedDevices field. When the design on how to support node level best effort topology aware scheduling is sorted out, perhaps we can extend the struct to allow Kubelet to make its local decisions. 4) The DeviceIDs of a ComputeResource are either set by device plugins for extended resources or by Kubelet when we also represent primary resources through the new resource API in the future. They can map to different compute resource units through the DeviceUnits field that we may add later. E.g., for CPU, quite likely we will export a single vCPU device with X units, unless the node is configured to also export other topology information like numa. I think that should be controlled by a kubelet config to match the underlying node cpu conditions (e.g., depending on CPU hw properties or whether auto numa is enabled).

Reply:
@vikaschoudhary16 and I just had another sync-up on this discussion. TBH, we have gone back and forth between the two models, i.e., having scheduler encode ResourceClass -> ComputeResource binding in ContainerSpec through ComputeResource name v.s. having scheduler encode ResourceClass -> ComputeResource binding in ContainerSpec through allocated deviceIds. The main pro for the first approach is that it requires less changes on the current scheduler and kubelet logic as we already implemented the device allocation logic in kubelet deviceManager. The main pro for the later is to allow easier device property upgrade and avoid introducing another API field. We can change back to the first model as proposed in the original google doc https://docs.google.com/document/d/1qlA2vZqeF4PTOz7yGcImj_k5gon6liraHYjxNAqUujc/#heading=h.3r9r49o2eag3 if people feel that is the right approach to start with. That means we will add a `Name` field in `ComputeResource` struct that can be generated through “nodeName-resourceName-propertyHash” naming convention, and Scheduler can record ResourceClass -> ComputeResource binding in ContainerSpec through that ComputeResource name. For simplicity, we can request that device property change requires a node drain to make sure no active pods would use an invalid ComputeResource name. This way, most of the deviceManager device allocation logic can still stay within kubelet.